docs/plugins.md

```diff
@@ -184,20 +236,162 @@ function embed(
+function preprocess(ast: AST, options: object): AST;
+```
+
+#### (optional) `insertPragma`
+
+A plugin can implement how a pragma comment is inserted in the resulting
+code when the `--insert-pragma` option is used, in the `insertPragma`
+function. Its signature is:
 
 ```ts
 function insertPragma(text: string): string;
 ```
 
-_(Optional)_ The preprocess function can process the ast from parser before passing into `print` function.
+#### Handling comments in a printer
+
+Comments are often not part of a language's AST and present a challenge
+for pretty printers. A Prettier plugin can either print comments itself
+in its `print` function or rely on Prettier's comment algorithm.
+
+By default, if the AST has a top-level `comments` attribute,
+Prettier assumes that `comments` stores an array of comment nodes.
+Prettier will then use the provided `parsers[<plugin>].locStart`/`locEnd`
+functions to search for the AST node that each comment "belongs" to.
+Comments are then attached to these nodes **mutating the AST in the
+process**. The `*Comment` functions are used to adjust Prettier's
+algorithm. Once attached to the AST, Prettier will automatically call
+the `printComment(path, options): Doc` function and insert the returned
+doc into the (hopefully) correct place.
+
+#### (optional) `printComment`
+
+Called whenever a comment node needs to be printed. It has the signature:
 
 ```ts
-function preprocess(ast: AST, options: object): AST;
+function printComment(
+  // Path to the current comment node
+  commentPath: FastPath,
+  // Print a node with the current printer
+  options: object
+): Doc;
+```
+
+#### (optional) `canAttachComment`
+
+```ts
+function canAttachComment(node: AST): boolean;
 ```
 
+This function is used for deciding whether a comment can be attached to
+a particular AST node. By default, _all_ AST properties are traversed
+searching for nodes that comments can be attached to. This function
+is used to prevent comments from being attached to a particular node.
+A typical implementation looks like
+
+```js
+function canAttachComment(node) {
+  return node.type && node.type !== "comment";
+}
+```
+
+#### (optional) `isBlockComment`
+
+```ts
+function canAttachComment(node: AST): boolean;
+```
+
+Returns whether or not the AST node is a block comment.
+
+#### (optional) `handleComments`
+
+The `handleComments` object contains three optional functions, each with signature
+
+```ts
+function(
+	// The AST node corresponding to the comment
+	comment: AST,
+	// The full source code text
+	text: string,
+	// The global options object
+	options: object,
+	// The AST
+	ast: AST,
+	// Whether this comment is the last comment
+	isLastComment: boolean
+): boolean
+```
+
+These functions are used to override Prettier's default comment
+assignment algorithm. `ownLine`/`endOfLine`/`remaining` is expected
+to either manually assign a comment to a node (removing it from the
+`astRoot.comments` array) and return `true`, or return a `falsy` value
```

Comment:
`handleComments.*` don't remove comments from that array.

The reference to that array is removed from the AST root, but this fact isn't directly related to the `handleComments` functions:

https://github.com/prettier/prettier/blob/8718bb4be44f6fd1aed8fcf91717f38931643dd7/src/main/core.js#L56-L60